servicetalk-encoding-api/src/main/java/io/servicetalk/encoding/api/AbstractZipContentCodec.java

```diff
@@ -113,67 +114,31 @@ public final Buffer encode(final Buffer src, final int offset, final int length,
 
                     @Override
                     public void onSubscribe(PublisherSource.Subscription subscription) {
-                        try {
-                            Buffer dst = allocator.newBuffer(chunkSize);
-                            stream = new SwappableBufferOutputStream(dst);
-                            // This will write header bytes on the stream, which will be consumed along with the first
-                            // onNext part
-                            output = newDeflaterOutputStream(stream);
-                        } catch (Exception e) {
-                            close(output);
-                            LOGGER.error("Error while encoding with {}", name(), e);
-                            deliverErrorFromSource(subscriber, e);
-                            return;
-                        }
-
                         subscriber.onSubscribe(subscription);
                     }
 
                     @Override
                     public void onNext(Buffer next) {
-                        assert output != null;
-                        assert stream != null;
-
-                        // onNext will produce AT-MOST N items (as received)
+                        // onNext will produce AT-MOST N items (from upstream)
                         // +1 for the encoding footer (ie. END_OF_STREAM)
                         try {
-                            if (next == END_OF_STREAM) {
-                                Buffer dst;
-                                if (!headerWritten) {
-                                    // Empty publisher will not produce any items other than the END_OF_STREAM
-                                    // so we need to keep the header & footer in the same chunk and deliver downstream.
-                                    dst = stream.buffer;
-                                } else {
-                                    // ZIP footer is 10 bytes
-                                    dst = allocator.newBuffer(10);
-                                    stream.swap(dst);
-                                }
+                            Buffer dst = allocator.newBuffer(next == END_OF_STREAM ? FOOTER_LEN : chunkSize);
+                            stream.swap(dst);
 
+                            if (!headerWritten) {
+                                // This will produce header bytes
+                                output = newDeflaterOutputStream(stream);
```

Comment:
can `output` be created up front too? I know this is `Closable`, but the existing approach doesn't account for `Subscription.cancel()` anyways and I would propose we revisit lifetime in a followup PR (for better isolation of changes).

Reply:
If output is created up front, then we need to allocate up-front too because it starts writing the header bytes immediately. This way we only allocate when we start consuming. I am re-working the overall lifetime + enhancements in a follow PR, let's keep further changes out of this one. I can revert if you prefer the old way, but I think this follows the 'cold' semantics of ST.